src/backend/mmap.rs

```diff
@@ -0,0 +1,151 @@
+extern crate memmap;
+extern crate failure;
+
+use self::failure::ResultExt;
+
+use super::Backend;
+
+use ::error;
+use ::std::io;
+use ::std::cmp;
+
+#[derive(Debug)]
+struct Mmap {
+    inner: memmap::MmapMut,
+    //End of data
+    pub end: usize,
+    //Mmap total len
+    pub len: usize
+}
+
+impl Mmap {
+    fn new(len: usize) -> io::Result<Self> {
+        let inner = memmap::MmapOptions::new().len(len).map_anon()?;
+
+        Ok(Self {
+            inner,
+            end: 0,
+            len
+        })
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        &self.inner[..self.end]
+    }
+
+    fn as_mut_slice(&mut self) -> &mut [u8] {
+        &mut self.inner[..self.end]
+    }
+
+    //Copies data to mmap and modifies data's end cursor.
+    fn write(&mut self, data: &[u8]) {
+        assert!(data.len() <= self.len);
+        self.end = data.len();
+        self.as_mut_slice().copy_from_slice(data);
+    }
+
+    #[inline]
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner.flush()
+    }
+
+    //Increases mmap size by max(old_size*2, new_size)
+    //Note that it doesn't copy original data
+    fn extend(&mut self, new_size: usize) -> io::Result<()> {
```

Comment:
Consider renaming this to `resize_no_copy` or similar, since extend usually implies that new data is appended, and that old data is not lost.